plugins/settings/settings_plugin.c

```diff
@@ -180,53 +266,209 @@ settings_plugin_get_boolean(
     gboolean val = g_key_file_get_boolean(config, SETTINGS_GROUP, key, &error);
 
     if (error) {
-        g_error_free(error);
-        /* Default */
-        return defval;
-    } else {
-        return val;
+        g_clear_error(&error);
+        val = g_key_file_get_boolean(self->defaults, SETTINGS_GROUP, key,
+            &error);
+        if (error) {
+            g_error_free(error);
+            /* Default */
+            return defval;
+        }
     }
+    return val;
 }
 
 static
 gboolean
 settings_plugin_nfc_enabled(
+    SettingsPlugin* self,
     GKeyFile* config)
 {
-    return settings_plugin_get_boolean(config, SETTINGS_KEY_ENABLED, TRUE);
+    return settings_plugin_get_boolean(self, config,
+        SETTINGS_KEY_ENABLED, TRUE);
 }
 
 static
 gboolean
 settings_plugin_nfc_always_on(
+    SettingsPlugin* self,
     GKeyFile* config)
 {
-    return settings_plugin_get_boolean(config, SETTINGS_KEY_ALWAYS_ON, FALSE);
+    return settings_plugin_get_boolean(self, config,
+        SETTINGS_KEY_ALWAYS_ON, FALSE);
 }
 
 static
-void
-settings_plugin_update_config(
-    SettingsPlugin* self)
+gboolean
+settings_plugin_update_boolean(
+    SettingsPlugin* self,
+    GKeyFile* config,
+    const char* key,
+    gboolean value)
 {
-    GKeyFile* config = settings_plugin_load_config(self);
-    const gboolean enabled = settings_plugin_nfc_enabled(config);
+    const char* group = SETTINGS_GROUP;
+    GError* error = NULL;
+    gboolean default_value = g_key_file_get_boolean(self->defaults, group,
+        key, &error);
+    gboolean have_default = !error;
+    gboolean config_value;
+
+    g_clear_error(&error);
+    config_value = g_key_file_get_boolean(config, group, key, &error);
+    if (error) {
+        g_error_free(error);
+        /* Save it only if it's not the default value */
+        if (!have_default || value != default_value) {
+            g_key_file_set_boolean(config, group, key, value);
+            return TRUE;
+        }
+    } else if (have_default && value == default_value) {
+        /* Remove the default value from the config */
+        g_key_file_remove_key(config, group, key, NULL);
+        return TRUE;
+    } else if (config_value != value) {
```

Comment:
nit / matter of taste: having compares the same way around would be nice i.e. above has value == x vs  x != value here

Reply:
Matter of taste, indeed. I think when I was writing this, I was trying to make the code more compact and at the same time not to miss any combination of `value`, `config_value` and `default_value`. That's what I came up with 🤷‍♂️